build/xtask/src/sizes.rs

```diff
@@ -0,0 +1,158 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::convert::TryInto;
+use std::path::{Path, PathBuf};
+
+use anyhow::bail;
+use goblin::Object;
+use indexmap::IndexMap;
+
+use crate::Config;
+
+fn armv7m_suggest(size: u64) -> u64 {
+    // Nearest power of two
+    let mut i = 1;
+    while i < size {
+        i *= 2;
+    }
+    i
+}
+fn armv8m_suggest(size: u64) -> u64 {
+    // Nearest chunk of 32
+    ((size + 31) / 32) * 32
+}
+
+pub fn run(cfg: &Path) -> anyhow::Result<()> {
+    let cfg_contents = std::fs::read(&cfg)?;
+    let toml: Config = toml::from_slice(&cfg_contents)?;
+
+    let mut dist_dir = PathBuf::from("target");
+    dist_dir.push(&toml.name);
+    dist_dir.push("dist");
+
+    let mut memories = IndexMap::new();
+    for (name, out) in &toml.outputs {
+        // This is called after `dist`, which logs this overflow gracefully
+        let end = out.address.checked_add(out.size).unwrap();
+        memories.insert(name.clone(), out.address..end);
+    }
+
+    let output_region = |vaddr: u64| {
+        memories
+            .iter()
+            .find(|(_, region)| region.contains(&vaddr.try_into().unwrap()))
+            .map(|(name, _)| name.as_str())
+    };
+
+    struct Task<'a> {
+        name: &'a str,
+        stacksize: u32,
+        requires: &'a IndexMap<String, u32>,
+    }
+    let mut tasks: Vec<Task> = toml
+        .tasks
+        .iter()
+        .map(|(name, task)| Task {
+            name,
+            stacksize: task
+                .stacksize
+                .unwrap_or_else(|| toml.stacksize.unwrap()),
+            requires: &task.requires,
+        })
+        .collect();
+    tasks.push(Task {
+        name: "kernel",
+        stacksize: toml.stacksize.unwrap(),
+        requires: &toml.kernel.requires,
+    });
+
+    let suggest = if toml.target.starts_with("thumbv7m")
+        || toml.target.starts_with("thumbv7em")
+    {
+        armv7m_suggest
+    } else if toml.target.starts_with("thumbv8m") {
+        armv8m_suggest
+    } else {
+        panic!("Unknown target {}", toml.target);
+    };
+
+    let mut suggestions = Vec::new();
+    let mut first = false; // used to print a newline between tasks
```

Comment:
I believe your use of this variable is correct, but its name is inverted. It gets set on all _but_ first.

Reply:
Fixed (by removing it as part of other refactoring)